### PR TITLE
Default to the CR's namespace if the locked resources do not have one

### LINKED ISSUE #15: add default namespaces for resources with missing namespaces
if a resource is namespaced and its `metadata.namespace` is missing, default it to the namespace in which the resourcelocker CR is defined.

### PR DESCRIPTION
**Pull request by komish**
This change will cause the operator to inject a namespace in a `resourcelocker.spec.resources[*].object.metadata` if one is not provided at the CR. This gives users a bit more flexibility. Should also help facilitate some changes in the `alm-examples` for the operator's `ClusterServiceVersion` such that we can provide an example that will successfully install. 

Looking at the code, the `IsInitialized()` function seemed like the logic place to add this as it's already doing some defaulting/injection at the CRD for things like `patchType` and `serviceAccountRef`.

In practice, this takes a CR that is submitted to the API (while scoped to the default namespace) like this (note the missing exclusions, metadata namespace):

```yaml
apiVersion: redhatcop.redhat.io/v1alpha1
kind: ResourceLocker
metadata:
  name: locked-configmap-foo-bar-configmap
spec:
  resources:
    - object:
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: foo-bar-configmap
        data:
          foo: bar
  serviceAccountRef:
    name: default
```

And gives us the following result

```yaml
apiVersion: redhatcop.redhat.io/v1alpha1
kind: ResourceLocker
metadata:
  finalizers:
  - resourcelocker-controller
  generation: 1
  name: locked-configmap-foo-bar-configmap
  selfLink: /apis/redhatcop.redhat.io/v1alpha1/namespaces/default/resourcelockers/locked-configmap-foo-bar-configmap
spec:
  resources:
  - excludedPaths:
    - .spec.replicas
    - .metadata
    - .status
    object:
      apiVersion: v1
      data:
        foo: bar
      kind: ConfigMap
      metadata:
        name: foo-bar-configmap
        namespace: default
  serviceAccountRef:
    name: default
```

Fixes #15 

Also, this partially fixes [this](https://bugzilla.redhat.com/show_bug.cgi?id=1794821), but we would also need to update the [community-operators](https://github.com/operator-framework/community-operators/blob/master/community-operators/resource-locker-operator/0.1.2/resource-locker-operator.v0.1.2.clusterserviceversion.yaml#L5) repo to adjust the first alm-example. I can do some further testing there after this is merged and pushed.

### CODE REVIEW

File: pkg/controller/resourcelocker/resourcelocker_controller.go [hunk 1]
```diff
@@ -274,20 +274,65 @@ func getLockedResources(instance *redhatcopv1alpha1.ResourceLocker) ([]unstructu
...
+			// TODO: Should we use ManageError here?
+			log.Error(err, "Unable to get locked resource for object: ", instance.Spec.Resources[i].Object)
+		} else {
+			if lockedResource.GetNamespace() == "" {
+				// The value of the namespace is blank, so we need to set it. We only drop into
+				// this logic if we actually needed to make change.
+				log.Info("Locked object found without namespace. Injecting namespace for object.",
+					"Object.Metadata.Name", lockedResource.GetName(),
+					"NamespaceToSet", instance.Namespace)
+				lockedResource.SetNamespace(instance.Namespace)
```
Comment (raffaelespazzoli) [inline pkg/controller/resourcelocker/resourcelocker_controller.go:316]: before setting the namespace, please add a call to find out whether the resource is namespaced or not. Something like this:
```golang
func isNamespaced(gvk schema.GroupVersionKind, cfg *rest.Config) (bool, error) {
	discoveryClient, err := discovery.NewDiscoveryClientForConfig(cfg)
	if err != nil {
		log.Error(err, "Unable to get discovery client")
		return false, err
	}
	resourceList, err := discoveryClient.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
	if err != nil {
		log.Error(err, "Unable to get resource list for", "apiversion", gvk.GroupVersion().String())
		return false, err
	}
	for _, apiResource := range resourceList.APIResources {
		if apiResource.Kind == gvk.Kind {
			return apiResource.Namespaced, nil
		}
	}
	return false, errors.New("Unable to find type: " + gvk.String() + " in server")
}
```

**Comment by komish** [comment]
Good catch @raffaelespazzoli on checking whether the resource is a namespaced resource or not! I've added the logic now and the operator will spin up a discovery client to check if the resource is namespaced only when the `.metadata.namespace` key is empty.

Given the following input CR which has two objects, one of which is namespaced (ConfigMap):

```
apiVersion: redhatcop.redhat.io/v1alpha1
kind: ResourceLocker
metadata:
  name: multi-object-test
spec:
  resources:
    - object:
        apiVersion: v1
        kind: ConfigMap
        metadata:
          name: foo-bar-configmap
        data:
          foo: bar
    - object:
        apiVersion: rbac.authorization.k8s.io/v1
        kind: ClusterRole
        metadata:
          name: resource-locker-getter
        rules:
          - apiGroups:
            - resourcelockers.redhatcop.redhat.io
            resources:
            - resourcelockers
            verbs:
            - get
  serviceAccountRef:
    name: default
```

We see the following resulting `.spec` on the persisted object:

```
spec:
  resources:
  - excludedPaths:
    - .spec.replicas
    - .metadata
    - .status
    object:
      apiVersion: v1
      data:
        foo: bar
      kind: ConfigMap
      metadata:     # namespace injected here!
        name: foo-bar-configmap
        namespace: default
  - excludedPaths:
    - .metadata
    - .status
    - .spec.replicas
    object:
      apiVersion: rbac.authorization.k8s.io/v1
      kind: ClusterRole
      metadata:     # nothing injected here!
        name: resource-locker-getter
      rules:
      - apiGroups:
        - resourcelockers.redhatcop.redhat.io
        resources:
        - resourcelockers
        verbs:
        - get
  serviceAccountRef:
    name: default
```

**Comment by komish** [comment]
@raffaelespazzoli Quick bump for review!

File: pkg/controller/resourcelocker/resourcelocker_controller.go [hunk 2]
```diff
@@ -274,20 +274,77 @@ func getLockedResources(instance *redhatcopv1alpha1.ResourceLocker) ([]unstructu
...
+				if err != nil {
+					// we were unable to get info from the server telling us the resource was namespaced.
+					log.Error(err, "Unable to determine if locked resource is namespaced", "object", instance.Spec.Resources[i].Object)
+				}
+			}
+
+			if isNamespaced {
+				// the resource is namespaced and is missing a namespace value, so we need
+				// to inject one.
+				log.Info("Locked object found without namespace. Injecting namespace for namespaced object.",
```
Comment (raffaelespazzoli) [inline pkg/controller/resourcelocker/resourcelocker_controller.go:325]: if this is useful to keep please change it to log.V(1).Info(...)

File: pkg/controller/resourcelocker/resourcelocker_controller.go [hunk 3]
```diff
@@ -302,9 +359,40 @@ func (r *ReconcileResourceLocker) IsInitialized(instance *redhatcopv1alpha1.Reso
...
 }
 
+// isNamespaced returns true if the provided resource is namespaced.
+func (r *ReconcileResourceLocker) isNamespaced(object *unstructured.Unstructured) (bool, error) {
+	var namespaced bool
+
+	// establish the discovery client
+	dc, err := r.GetDiscoveryClient()
+	if err != nil {
+		return namespaced, err
```
Comment (raffaelespazzoli) [inline pkg/controller/resourcelocker/resourcelocker_controller.go:373]: please add a log.Error statement here

File: pkg/controller/resourcelocker/resourcelocker_controller.go [hunk 4]
```diff
@@ -302,9 +359,40 @@ func (r *ReconcileResourceLocker) IsInitialized(instance *redhatcopv1alpha1.Reso
...
+	}
+
+	// get gvk for the unstructured object
+	gvk := object.GroupVersionKind()
+
+	// get the resources for the relevant gv
+	resourceList, err := dc.ServerResourcesForGroupVersion(gvk.GroupVersion().String())
+	if err != nil {
+		return namespaced, err
```
Comment (raffaelespazzoli) [inline pkg/controller/resourcelocker/resourcelocker_controller.go:382]: please add a log.Error statement here

**Comment by raffaelespazzoli** [comment]
sorry for the additional changes, after these we should be good.
Also one question, what kind of test are you running on this?

**Comment by komish** [comment]
> sorry for the additional changes, after these we should be good.
> Also one question, what kind of test are you running on this?

I've only been testing to see if the proper mutations happen on the CR as I expect, but I'm now seeing a few tests in your `test/` directory so I'll try those. Will submit remaining changes soon!

File: pkg/controller/resourcelocker/resourcelocker_controller.go (hunk 2 above)
Comment (komish) [reply pkg/controller/resourcelocker/resourcelocker_controller.go]: Done!

File: pkg/controller/resourcelocker/resourcelocker_controller.go (hunk 3 above)
Comment (komish) [reply pkg/controller/resourcelocker/resourcelocker_controller.go:375]: Done!

File: pkg/controller/resourcelocker/resourcelocker_controller.go (hunk 4 above)
Comment (komish) [reply pkg/controller/resourcelocker/resourcelocker_controller.go:388]: Done!

**Comment by komish** [comment]
@raffaelespazzoli latest changes have been implemented, thanks for the suggestions!

I ran through the tests in the `test/` directory and things created without issue. I also stripped the namespace in the `resource_test.yaml` file and it was injected as expected.

Lastly, looking at the multi-object example I included previously, this creates without issue and namespace injections happen where I expect.

```yaml
...
  spec:
    resources:
    - excludedPaths:
      - .metadata
      - .status
      - .spec.replicas
      object:
        apiVersion: v1
        data:
          foo: bar
        kind: ConfigMap
        metadata:           #namespace injected here!
          name: foo-bar-configmap
          namespace: resource-locker-test 
    - excludedPaths:
      - .metadata
      - .status
      - .spec.replicas
      object:
        apiVersion: rbac.authorization.k8s.io/v1
        kind: ClusterRole
        metadata:           #namespace not injected here
          name: resource-locker-getter
        rules:
        - apiGroups:
          - resourcelockers.redhatcop.redhat.io
          resources:
          - resourcelockers
          verbs:
          - get
    serviceAccountRef:
...
```

I do see what I think is a bug in the patch maintainer logic, but this appears to be unrelated. Including here in case you want to open a separate bug about this. Probably can be tested against ClusterRole objects.

```json
{
  "level": "error",
  "ts": 1588013970.162233,
  "logger": "controller_lockedresource",
  "msg": "unable to apply",
  "patch": "W3sib3AiOiJyZW1vdmUiLCJwYXRoIjoiL3NwZWMvcmVwbGljYXMifV0=",
  "to json": "{\"apiVersion\":\"rbac.authorization.k8s.io/v1\",\"kind\":\"ClusterRole\",\"rules\":[{\"apiGroups\":[\"resourcelockers.redhatcop.redhat.io\"],\"resources\":[\"resourcelockers\"],\"verbs\":[\"get\"]}]}",
  "error": "remove operation does not apply: doc is missing path: \"/spec/replicas\": missing value",
  "errorVerbose": "missing value\ngithub.com/evanphx/json-patch.init\n\t/Users/myusername/.go/pkg/mod/github.com/evanphx/json-patch@v4.5.0+incompatible/patch.go:31\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5222\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5217\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5217\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5217\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5217\nruntime.doInit\n\t/usr/local/go/src/runtime/proc.go:5217\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:190\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357\nremove operation does not apply: doc is missing path: \"/spec/replicas\"\ngithub.com/evanphx/json-patch.Patch.remove\n\t/Users/myusername/.go/pkg/mod/github.com/evanphx/json-patch@v4.5.0+incompatible/patch.go:525\ngithub.com/evanphx/json-patch.Patch.ApplyIndent\n\t/Users/myusername/.go/pkg/mod/github.com/evanphx/json-patch@v4.5.0+incompatible/patch.go:738\ngithub.com/evanphx/json-patch.Patch.Apply\n\t/Users/myusername/.go/pkg/mod/github.com/evanphx/json-patch@v4.5.0+incompatible/patch.go:710\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.filterOutPaths\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/patch.go:29\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.(*LockedResourceReconciler).isEqual\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/resource-reconciler.go:222\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.(*LockedResourceReconciler).Reconcile\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/resource-reconciler.go:150\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1357",
  "stacktrace": "github.com/go-logr/zapr.(*zapLogger).Error\n\t/Users/myusername/.go/pkg/mod/github.com/go-logr/zapr@v0.1.1/zapr.go:128\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.filterOutPaths\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/patch.go:34\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.(*LockedResourceReconciler).isEqual\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/resource-reconciler.go:222\ngithub.com/redhat-cop/resource-locker-operator/pkg/lockedresourcecontroller.(*LockedResourceReconciler).Reconcile\n\t/Users/myusername/Projects/rlocker/resource-locker-operator/pkg/lockedresourcecontroller/resource-reconciler.go:150\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:256\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:232\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker\n\t/Users/myusername/.go/pkg/mod/sigs.k8s.io/controller-runtime@v0.4.0/pkg/internal/controller/controller.go:211\nk8s.io/apimachinery/pkg/util/wait.JitterUntil.func1\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:152\nk8s.io/apimachinery/pkg/util/wait.JitterUntil\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:153\nk8s.io/apimachinery/pkg/util/wait.Until\n\t/Users/myusername/.go/pkg/mod/k8s.io/apimachinery@v0.0.0-20191004115801-a2eda9f80ab8/pkg/util/wait/wait.go:88"
}
```

Once this is ready to merged, I would be interesting in knowing the timeline for getting a new bundle pushed to https://github.com/operator-framework/community-operators/ - I would like to make one final change to the CSV's `alm-examples` to resolve the referenced Bugzilla bug listed in the related issue. Would love to see about getting that issue resolved by the end of this week if possible!

**Pull request merged**